core/store/src/flat/manager.rs

```diff
@@ -93,6 +96,107 @@ impl FlatStorageManager {
         Ok(())
     }
 
+    /// Update flat storage for given processed or caught up block, which includes:
+    /// - merge deltas from current flat storage head to new one;
+    /// - update flat storage head to the hash of final block visible from given one;
+    /// - remove info about unreachable blocks from memory.
+    pub fn update_flat_storage_for_shard(
+        &self,
+        shard_uid: ShardUId,
+        block: &Block,
+    ) -> Result<(), StorageError> {
+        if let Some(flat_storage) = self.get_flat_storage_for_shard(shard_uid) {
+            let mut new_flat_head = *block.header().last_final_block();
+            if new_flat_head == CryptoHash::default() {
+                let genesis_hash = get_genesis_hash(&self.0.store)
```

Comment:
Note: This part is changed from the original implementation. This is expected to be super rare but it makes sense to get the genesis hash from store misc column instead of the stored value in chain.